### PR TITLE
[SYCL][NFC] Fix warnings coming out of SYCL headers.

### PR DESCRIPTION
**Pull request by AlexeySachkov**
The problem was discovered in intel/llvm#3777. Fixed by wrapping `x`
`_SYCL_SPAN_ASSERT` argument into parentheses.

Example of the warning/error:
```
sycl/CL/sycl/sycl_span.hpp:515:82: error: '&&' within '||' [-Werror,-Wlogical-op-parentheses]
    (static_cast <bool> ((_Count == dynamic_extent || _Count <= size() - _Offset && "Offset + count out of range in span::subspan()"))·
? void (0) : __assert_fail ("(_Count == dynamic_extent || _Count <= size() - _Offset && \"Offset + count out of range in span::subspan(
)\")", "/sycl/CL/sycl/sycl_span.hpp", 516, __extension__ __PRETTY_FUNCTION__));

sycl/CL/sycl/sycl_span.hpp:515:82: note: place parentheses around the '&&' expression· to silence this warning
```

### CODE REVIEW

**Review by dm-vodopyanov — APPROVED** [review]
LGTM. @bader, can you please merge it today when it will be convenient to unblock #3777?

**Comment by bader** [comment]
@AlexeySachkov, @dm-vodopyanov, I'm okay to merge this PR as soon as pre-commit tests pass, but ideally, the test case triggering the issue should be added to this PR.

**Comment by AlexeySachkov** [comment]
> @AlexeySachkov, @dm-vodopyanov, I'm okay to merge this PR as soon as pre-commit tests pass, but ideally, the test case triggering the issue should be added to this PR.

The test which triggered the issue is existing `warnings.cpp`: [logs](http://ci.llvm.intel.com:8010/#/builders/2/builds/14842), the issue is only reproducible when integration footer is involved. I will update the test

**Comment by AlexeySachkov** [comment]
> I will update the test

Oops, too soon. The integration footer is not useable yet without #3777 and #3777 encounters this issue. I suggest that we merge this one as-is

**Comment by bader** [comment]
You could write a stand-alone reproducer calling `_SYCL_SPAN_ASSERT` with an expression passed to the first argument triggering the warning and tread it as an error. I don't think there should be dependency on integration footer or #3777.

**Comment by AlexeySachkov** [comment]
> You could write a stand-alone reproducer calling `_SYCL_SPAN_ASSERT` with an expression passed to the first argument triggering the warning and tread it as an error. I don't think there should be dependency on integration footer or #3777.

I'm not exactly sure that this is a reliable thing, because I would expect `_SYCL_SPAN_ASSERT` to be undefined at the end of the file - it is not the case right now, but it is the case for a lot of internal marco. For example, the same problem with parentheses exists in `types.hpp` in `__SYCL_GENERATE_CONVERT_IMPL` macro - I won't be able to write isolated test for it, because the macro is undefined after its use.

Problem with `types.hpp` was found by manually launching the preprocessor first and then launching compiler on a preprocessed file - I don't know why we don't see those warnings if we go through a single-command compilation

**Comment by bader** [comment]
> Problem with `types.hpp` was found by manually launching the preprocessor first and then launching compiler on a preprocessed file - I don't know why we don't see those warnings if we go through a single-command compilation

I assume it's because SYCL headers are handled as system/compiler headers...

**Comment by bader** [comment]
Did you try to build the test in non-SYCL mode (i.e. w/o `-fsycl` flag)?

**Comment by AlexeySachkov** [comment]
> Did you try to build the test in non-SYCL mode (i.e. w/o `-fsycl` flag)?

Not yet, but it seems that I was able to reproduce the original problem by explicitly instantiating some SYCL classes in the test

**Comment by bader** [comment]
Jenkins/Precommit is still running, but I see failed tests already.

**Comment by AlexeySachkov** [comment]
> > Problem with `types.hpp` was found by manually launching the preprocessor first and then launching compiler on a preprocessed file - I don't know why we don't see those warnings if we go through a single-command compilation
> 
> I assume it's because SYCL headers are handled as system/compiler headers...

`--no-system-header-prefix=CL/sycl` should have prevented that, but seems like it doesn't work in all cases

**Comment by AlexeySachkov** [comment]
> Did you try to build the test in non-SYCL mode (i.e. w/o `-fsycl` flag)?

@bader, I've tried this and it doesn't catch errors which are visible with manual preprocessing + compilation of preprocessed file steps

**Comment by bader** [comment]
> Not yet, but it seems that I was able to reproduce the original problem by explicitly instantiating some SYCL classes in the test

I think the problem with reproducing the issue might be that the problem is in the template which is not instantiated. When the code is compiled, the both proposed ways should work (i.e. `-fsycl` + `--no-system-header-prefix=CL/sycl` and no `-fsycl` + `-I/path/to/sycl/headers`).

**Comment by AlexeySachkov** [comment]
> > Not yet, but it seems that I was able to reproduce the original problem by explicitly instantiating some SYCL classes in the test
> 
> I think the problem with reproducing the issue might be that the problem is in the template which is not instantiated. When the code is compiled, the both proposed ways should work (i.e. `-fsycl` + `--no-system-header-prefix=CL/sycl` and no `-fsycl` + `-I/path/to/sycl/headers`).

Still can't reproduce them with the following commands:
```
// RUN: %clangxx -fsycl --no-system-header-prefix=CL/sycl -fsyntax-only -Wall -Wextra -Wno-ignored-attributes -Wno-deprecated-declarations -Wpessimizing-move -Wunused-variable -Wmismatched-tags -Wunneeded-internal-declaration -Werror -Wno-unknown-cuda-version -Wno-unused-parameter %s -o %t.out                                                                                             
// RUN: %clangxx -std=c++17 -I %sycl_include -I %sycl_include/sycl -fsyntax-only -Wall -Wextra -Wno-ignored-attributes -Wno-deprecated-declarations -Wpessimizing-move -Wunused-variable -Wmismatched-tags -Wunneeded-internal-declaration -Werror -Wno-unknown-cuda-version -Wno-unused-parameter %s -o %t.out                                                                                    
```

File: sycl/test/warnings/warnings.cpp [hunk 1]
```diff
@@ -1,17 +1,42 @@
 // RUN: %clangxx -fsycl --no-system-header-prefix=CL/sycl -fsyntax-only -Wall -Wextra -Wno-ignored-attributes -Wno-deprecated-declarations -Wpessimizing-move -Wunused-variable -Wmismatched-tags -Wunneeded-internal-declaration -Werror -Wno-unknown-cuda-version -Wno-unused-parameter %s -o %t.out
-
+// RUN: %clangxx -fsycl -E --no-system-header-prefix=CL/sycl %s -o %t.ii
+// RUN: %clangxx -fsycl -fsyntax-only -Wall -Wextra -Wno-ignored-attributes -Wno-deprecated-declarations -Wpessimizing-move -Wunused-variable -Wmismatched-tags -Wunneeded-internal-declaration -Werror -Wno-unknown-cuda-version -Wno-unused-parameter %t.ii -o %t.out
```
Comment (bader) [inline sycl/test/warnings/warnings.cpp:3]:
```suggestion
// RUN: %clangxx -fsycl -fsyntax-only -Wall -Wextra -Werror -Wno-ignored-attributes -Wno-deprecated-declarations -Wpessimizing-move -Wunused-variable -Wmismatched-tags -Wunneeded-internal-declaration -Wno-unknown-cuda-version -Wno-unused-parameter %t.ii
```

**Comment by bader** [comment]
> > > Not yet, but it seems that I was able to reproduce the original problem by explicitly instantiating some SYCL classes in the test
> > 
> > 
> > I think the problem with reproducing the issue might be that the problem is in the template which is not instantiated. When the code is compiled, the both proposed ways should work (i.e. `-fsycl` + `--no-system-header-prefix=CL/sycl` and no `-fsycl` + `-I/path/to/sycl/headers`).
> 
> Still can't reproduce them with the following commands:
> 
> ```
> // RUN: %clangxx -fsycl --no-system-header-prefix=CL/sycl -fsyntax-only -Wall -Wextra -Wno-ignored-attributes -Wno-deprecated-declarations -Wpessimizing-move -Wunused-variable -Wmismatched-tags -Wunneeded-internal-declaration -Werror -Wno-unknown-cuda-version -Wno-unused-parameter %s -o %t.out                                                                                             
> // RUN: %clangxx -std=c++17 -I %sycl_include -I %sycl_include/sycl -fsyntax-only -Wall -Wextra -Wno-ignored-attributes -Wno-deprecated-declarations -Wpessimizing-move -Wunused-variable -Wmismatched-tags -Wunneeded-internal-declaration -Werror -Wno-unknown-cuda-version -Wno-unused-parameter %s -o %t.out                                                                                    
> ```

Interesting... can you dump the driver actions and compare with the actions from your reproducer?

File: sycl/test/warnings/warnings.cpp (hunk 1 above)
Comment (AlexeySachkov) [inline sycl/test/warnings/warnings.cpp:3]: BTW, I don't know why `-Wno-unused-parameter` was added here, but we can pass this test just fine without disabling that diagnostic

**Comment by AlexeySachkov** [comment]
> Interesting... can you dump the driver actions and compare with the actions from your reproducer?

From what I see, both `clang++ -fsycl` and just `clang++` do not launch the preprocessor as a separate step, i.e. there are no separate `clang` invocation with `-E`

File: sycl/test/warnings/warnings.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply sycl/test/warnings/warnings.cpp]: Applied in b16be58

**Review by bader — APPROVED** [review]

**Review by dm-vodopyanov — APPROVED** [review]

**Review by turinevgeny — APPROVED** [review]

File: sycl/test/warnings/warnings.cpp (hunk 1 above)
Comment (AlexeySachkov) [reply sycl/test/warnings/warnings.cpp]: Outlined this into #3980

**Pull request merged**